### PR TITLE
Equal or null optimization

### PR DESCRIPTION
**Pull request by PedroTadim**
Hello, Duckies! Here is my very first contribution to the Duck! However, I have some questions before the reviews.

This pull request implements the a=b OR (a IS NULL AND b IS NULL) to a IS NOT DISTINCT FROM b rewrite as part of issue #3509.

1. From writing SQL optimizers in the past I found some corner cases such as equality between two different types, (eg INT and BIGINT), and an upcast is needed. This creates corner cases for optimizers such as this one where the expressions won't match between both sides of the OR due to the upcast under the equality. I wonder if such cases have been handled before because I didn't spot it from other optimizers.
2. Is there any documentation about writing optimizers? That's because it's not clear to me what is the role of the "Rule". I see some optimizers with rule clauses, but later during the optimization phase, they apply other conditions, which makes it confusing for me :(
3.  I see a new expression is always returned when the expression optimizer makes changes. but could we just overwrite the input expression to avoid allocations, or a single expression can have more than one reference?
4. In MonetDB there was a performance issue because there were multiple optimizers present, so we had to pack them together in single calls. I see many here: https://github.com/PedroTadim/duckdb/blob/master/src/optimizer/optimizer.cpp#L27 It would be desirable to pack them in the future? Same here: https://github.com/PedroTadim/duckdb/blob/master/src/optimizer/optimizer.cpp#L66

Thanks for your support!

### CODE REVIEW

File: test/sql/optimizer/expression/test_equal_or_null_optimization.test [hunk 1]
```diff
@@ -0,0 +1,35 @@
+# name: test/sql/optimizer/expression/test_equal_or_null_optimization.test
+# description: Test a=b OR (a IS NULL AND b IS NULL) to a IS NOT DISTINCT FROM b
+# group: [expression]
+
+statement ok
+PRAGMA enable_verification
+
+statement ok
```
Comment (hawkfish) [inline test/sql/optimizer/expression/test_equal_or_null_optimization.test:8]: Transactions aren't needed in tests (it is done automatically for you.)

File: test/sql/optimizer/expression/test_equal_or_null_optimization.test [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+START TRANSACTION;
+
+statement ok
+CREATE TABLE test(i INTEGER, j INTEGER);
+
+statement ok
+INSERT INTO test VALUES (1,1), (2,2), (NULL,NULL)
```
Comment (hawkfish) [inline test/sql/optimizer/expression/test_equal_or_null_optimization.test:15]: Maybe some values that fail the test?

File: test/sql/optimizer/expression/test_equal_or_null_optimization.test [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+NULL
+
+query I rowsort
+SELECT i FROM test WHERE (i IS NULL AND j IS NULL) OR (i=j);
+----
+1
+2
+NULL
+
+statement ok
```
Comment (hawkfish) [inline test/sql/optimizer/expression/test_equal_or_null_optimization.test:34]: Can we add some tests that cover the rejected cases?

File: src/optimizer/rule/equal_or_null_simplification.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+	}
+	return nullptr;
+}
+
+unique_ptr<Expression> EqualOrNullSimplification::Apply(LogicalOperator &op, vector<Expression *> &bindings,
+                                                        bool &changes_made, bool is_root) {
+	const auto or_exp = bindings[0];
+
+	if (((BoundConjunctionExpression *)or_exp)->children.size() == 2) {
+		const auto or_cast = (BoundConjunctionExpression *)or_exp;
```
Comment (hawkfish) [inline src/optimizer/rule/equal_or_null_simplification.cpp:55]: Another small readability nit: Can we just pull this up before the `if` and simplify the `if` test?

**Review by hawkfish — COMMENTED** [review]
Thanks for the PR!  These semantics were Tableau's, so I really appreciate the value of the optimisation.  The code looks correct to me, just some readability and coverage changes.

File: src/optimizer/rule/equal_or_null_simplification.cpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+		const auto left_exp = or_cast->children[0].get(), right_exp = or_cast->children[1].get();
+		// Test for: a=b OR (a IS NULL AND b IS NULL)
+		auto one_try = TryRewriteEqualOrIsNull(left_exp, right_exp);
+		if (one_try != nullptr) {
```
Comment (Mytherin) [inline src/optimizer/rule/equal_or_null_simplification.cpp:59]: `!one_try`

File: src/optimizer/rule/equal_or_null_simplification.cpp [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+	return nullptr;
+}
+
+unique_ptr<Expression> EqualOrNullSimplification::Apply(LogicalOperator &op, vector<Expression *> &bindings,
+                                                        bool &changes_made, bool is_root) {
+	const auto or_exp = bindings[0];
+
+	if (((BoundConjunctionExpression *)or_exp)->children.size() == 2) {
+		const auto or_cast = (BoundConjunctionExpression *)or_exp;
+		const auto left_exp = or_cast->children[0].get(), right_exp = or_cast->children[1].get();
```
Comment (Mytherin) [inline src/optimizer/rule/equal_or_null_simplification.cpp:56]: We generally avoid creating multiple variables on one line

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! Looks good. Here is some feedback from my side:

**Comment by Mytherin** [comment]
> From writing SQL optimizers in the past I found some corner cases such as equality between two different types, (eg INT and BIGINT), and an upcast is needed. This creates corner cases for optimizers such as this one where the expressions won't match between both sides of the OR due to the upcast under the equality. I wonder if such cases have been handled before because I didn't spot it from other optimizers.

There is no general infrastructure to handle this, this needs to be handled on a per-optimizer basis. It is certainly a good point and would be interesting to extend this and other optimizers to correctly handle this case.

> Is there any documentation about writing optimizers? That's because it's not clear to me what is the role of the "Rule". I see some optimizers with rule clauses, but later during the optimization phase, they apply other conditions, which makes it confusing for me :(

Rules are used for expression rewrites, the other optimizers do more complex operations on the tree itself rather than only on expressions, and there is no generic infrastructure for that (yet). 

> I see a new expression is always returned when the expression optimizer makes changes. but could we just overwrite the input expression to avoid allocations, or a single expression can have more than one reference?

You can return `nullptr` and no change will be made to the original expression

> In MonetDB there was a performance issue because there were multiple optimizers present, so we had to pack them together in single calls. I see many here: https://github.com/PedroTadim/duckdb/blob/master/src/optimizer/optimizer.cpp#L27 It would be desirable to pack them in the future? Same here: https://github.com/PedroTadim/duckdb/blob/master/src/optimizer/optimizer.cpp#L66

I agree that at some point we should spend some time on optimizing the optimizers - for now we have not encountered any situations where this has been a significant problem.

File: test/sql/optimizer/expression/test_equal_or_null_optimization.test [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+START TRANSACTION;
+
+statement ok
+CREATE TABLE test(i INTEGER, j INTEGER);
+
+statement ok
+INSERT INTO test VALUES (1,1), (2,2), (NULL,NULL)
+
+statement ok
+PRAGMA explain_output = 'OPTIMIZED_ONLY';
```
[the remark this answers is not in the document]
Comment (PedroTadim) [reply test/sql/optimizer/expression/test_equal_or_null_optimization.test:18]: I forgot to ask I would like to test a query plan with the EXPLAIN command, so the optimizer can be continuously tested? I could create a performance test instead with a join: hash join generated from IS DISTINCT FROM vs nested loop join from the unoptimized version generated by the OR clause.

File: src/optimizer/rule/equal_or_null_simplification.cpp [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+namespace duckdb {
+
+EqualOrNullSimplification::EqualOrNullSimplification(ExpressionRewriter &rewriter) : Rule(rewriter) {
+	// match on OR conjunction
+	root = make_unique<ExpressionMatcher>();
+	root->expr_type = make_unique<SpecificExpressionTypeMatcher>(ExpressionType::CONJUNCTION_OR);
+}
+
+// a=b OR (a IS NULL AND b IS NULL) to a IS NOT DISTINCT FROM b
+static unique_ptr<Expression> TryRewriteEqualOrIsNull(const Expression *e1, const Expression *e2) {
```
[the remark this answers is not in the document]
Comment (PedroTadim) [reply src/optimizer/rule/equal_or_null_simplification.cpp:16]: Yes they should be renamed

File: src/optimizer/rule/equal_or_null_simplification.cpp (hunk 4 above)
Comment (PedroTadim) [reply src/optimizer/rule/equal_or_null_simplification.cpp:55]: Yes, I can try the trick where I do the cast before the test. As long as the pointer is not dereferenced, it should be fine to do the cast even if the expression type is not the same.

File: src/optimizer/rule/equal_or_null_simplification.cpp (hunk 6 above)
Comment (PedroTadim) [reply src/optimizer/rule/equal_or_null_simplification.cpp:56]: Sure, fixed

File: src/optimizer/rule/equal_or_null_simplification.cpp [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+#include "duckdb/planner/expression/bound_comparison_expression.hpp"
+#include "duckdb/planner/expression/bound_conjunction_expression.hpp"
+#include "duckdb/planner/expression/bound_operator_expression.hpp"
+
+namespace duckdb {
+
+EqualOrNullSimplification::EqualOrNullSimplification(ExpressionRewriter &rewriter) : Rule(rewriter) {
+	// match on OR conjunction
+	root = make_unique<ExpressionMatcher>();
+	root->expr_type = make_unique<SpecificExpressionTypeMatcher>(ExpressionType::CONJUNCTION_OR);
```
[the remark this answers is not in the document]
Comment (PedroTadim) [reply src/optimizer/rule/equal_or_null_simplification.cpp:12]: I need help with this. I cannot make it work. For other optimizers I see they checking for matches during the apply phase. This becomes confusing.

[the remark this answers is not in the document]
Comment (PedroTadim) [reply src/optimizer/rule/equal_or_null_simplification.cpp:12]: I got it now

File: src/optimizer/rule/equal_or_null_simplification.cpp [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+// a=b OR (a IS NULL AND b IS NULL) to a IS NOT DISTINCT FROM b
+static unique_ptr<Expression> TryRewriteEqualOrIsNull(const Expression *e1, const Expression *e2) {
+	if (e1->type == ExpressionType::COMPARE_EQUAL && e2->type == ExpressionType::CONJUNCTION_AND &&
```
[the remark this answers is not in the document]
Comment (PedroTadim) [reply src/optimizer/rule/equal_or_null_simplification.cpp:17]: Fixed

File: src/optimizer/rule/equal_or_null_simplification.cpp [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+		}
+	}
+	return nullptr;
+}
+
+unique_ptr<Expression> EqualOrNullSimplification::Apply(LogicalOperator &op, vector<Expression *> &bindings,
+                                                        bool &changes_made, bool is_root) {
+	const auto or_exp = bindings[0];
+
+	if (((BoundConjunctionExpression *)or_exp)->children.size() == 2) {
```
[the remark this answers is not in the document]
Comment (PedroTadim) [reply src/optimizer/rule/equal_or_null_simplification.cpp:54]: Done

File: test/sql/optimizer/expression/test_equal_or_null_optimization.test [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+
+query I rowsort
+SELECT i FROM test WHERE (i=j) OR (i IS NULL AND j IS NULL);
```
[the remark this answers is not in the document]
Comment (PedroTadim) [reply test/sql/optimizer/expression/test_equal_or_null_optimization.test:18]: Yes, I forgot to ask about this, my apologies.

File: test/sql/optimizer/expression/test_equal_or_null_optimization.test (hunk 3 above)
Comment (PedroTadim) [reply test/sql/optimizer/expression/test_equal_or_null_optimization.test]: Just added them. I looked on the wrong test directory :)

**Comment by PedroTadim** [comment]
Yesterday I added more test cases and did the review fixes. I think it's in good shape to merge.

**Comment by Mytherin** [comment]
Thanks for the updates! Looks good to me too.

**Pull request merged**